[BB-1337] Initial implementation

lgp171188 opened the pull request:
Testing instructions are in the `README.md` file.

pomegranited commented on README.md:
```diff
@@ -0,0 +1,54 @@
+This requires administrator-level access to the repository.
+* Use the `<ngrok URL>/pull-requests` as the payload URL and set the `Content Type` to `application/json`.
+* Remember to set a secret and to select only the 'Pull request' event.
+* Create a GitHub web token.
```
nit: Please mention that no particular permissions are required for a public repo, and list the permissions that are required for private repos.

pomegranited commented on README.md:
```diff
@@ -0,0 +1,54 @@
+GitHub Webhook setup
+====================
+
+* Use a tool like `ngrok` to expose the local development server on the internet.
```
Is the `ngrok` step here just to make the service accessible from the internet?  If so, I think it's sufficient to say, "Run the service on an internet accessible URL.  In development, you can use a tool like `ngrok` to expose your local development server."

pomegranited commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+Flask==1.0.3
+PyGithub==1.43.7
```
Missing one:
```
flask-mail==0.9.1
```

pomegranited commented on settings.py:
```diff
@@ -0,0 +1,18 @@
+        'recipients': ['nobody@example.com', ],
+        'subject': '[ADR] Pull request: {repo} #{number} {action}',
+    },
+}
```
Since we're going to want to run this service without changing this file, could you make the `WATCH_CONFIG` (or at least its parts) also configurable from the env?

Then, you can also test these overrides without requiring a `test_settings.py`, by [patching os.environ](https://stackoverflow.com/a/31583528/4302112): 
```python
k = mock.patch.dict(os.environ,{'mytemp':'mytemp'})
```

lgp171188 replied:
@pomegranited, I was planning to do that, but since I was wary of the timebox at that time, I decided against doing it. But since I have time and this shouldn't take a lot of time, I will fix this.

pomegranited replied:
Thank you!  Works great.

pomegranited commented on pr_watcher_notifier/github_api.py:
```diff
@@ -0,0 +1,92 @@
+                target = get_target_branch(pr)
+                previous_head = data['before']
+                for modified_file in get_comparison_file_names(repo, target, previous_head):
+                    if fnmatch(modified_file, 'docs/*'):
```
Bug: need use the configured pattern(s) here too.

Your [`test_pr_synchronized_but_already_notified`](https://github.com/open-craft/pr_watcher_notifier/pull/1/files#diff-ee7175cd85f3b81c0c9c02ebba5283c6R131) didn't catch this because your [tests also match the default `docs/*`](https://github.com/open-craft/pr_watcher_notifier/pull/1/files#diff-2b5bde3dd5ae5bc227010549100ba6a8R6), so maybe change that too?

lgp171188 replied:
@pomegranited, thanks for catching this!

pomegranited commented on pr_watcher_notifier/templates/email_body.txt:
```diff
@@ -0,0 +1,6 @@
+Hi,
+
+Pull request #{{ number }} ({{ pr_url }}) against the {{ repo }} repository has been {{ action }}.
```
Could you add (all) the [matching files](https://github.com/open-craft/pr_watcher_notifier/pull/1/files#diff-fb78b20e3c0e71c3361c7925c2cded42R79) from the webhook to the message template?

And add some text like, "This PR contains files matching {{ pattern }}."

lgp171188 replied:
The current code uses only the first match to make a decision because I thought that it will be efficient. Thinking about it now, I don't think the performance difference is going to matter unless we have thousands of files modified. So I'll make the change.

pomegranited commented on pr_watcher_notifier/github_api.py:
```diff
@@ -0,0 +1,92 @@
+                        break
+                if matched:
+                    break
+        if matched:
```
Worth adding a comment explaining why what this block of code achieves, something like "Avoid sending multiple notifications out for a given PR by checking the previous commit for files matching `pattern`."

pomegranited commented on README.md:
````diff
@@ -0,0 +1,54 @@
+* Verify that the unit tests run without errors and pass.
+
+```bash
+$ pytest
````
Could you make this `pytest --cov=pr_watcher_notifier` instead?  The current coverage (90%) is sufficient, but it's useful to see.

pomegranited commented on README.md:
```diff
@@ -0,0 +1,54 @@
+* Use a tool like `ngrok` to expose the local development server on the internet.
+* Create a webhook on the repository to be watched by following [these instructions](https://developer.github.com/webhooks/creating/).
+This requires administrator-level access to the repository.
+* Use the `<ngrok URL>/pull-requests` as the payload URL and set the `Content Type` to `application/json`.
```
...and then this doesn't need to say `ngrok`.

pomegranited commented on README.md:
````diff
@@ -0,0 +1,54 @@
+```bash
+$ export FLASK_APP=run.py
+$ export FLASK_DEBUG=1
+$ flask run
````
To use the app on my non-localhost domain, I had to use `flask run --host=0.0.0.0`.  Is there any way to make that IP configurable from the settings or env?

lgp171188 replied:
@pomegranited, the `flask run` command doesn't read from settings or the environment variables. The old way of writing a file with `app.run(host='0.0.0.0')` in the `if __name__ == '__main__'` block of a script file is not recommended any more.

If there is time, I can create a script `run_dev.sh` reads the environment variable if it is set and modifies the command appropriately. Otherwise, we might have to stop with documenting this.

pomegranited replied:
👍

lgp171188 commented:
@pomegranited, I have updated the PR to address the review comments. Since I have added `PyYAML` to the requirements file, you will have to install it before testing.

nedbat commented on pr_watcher_notifier/github_api.py:
```diff
@@ -0,0 +1,101 @@
+    return pr.base.ref
+
+
+def should_send_notification(data):
```
A nit, but this function is not part of "the github api", it's got significant business logic in it.

lgp171188 replied:
@nedbat, you're right! I will move this to a more appropriate place in an update.

nedbat commented on pr_watcher_notifier/github_api.py:
```diff
@@ -0,0 +1,101 @@
+                    for pattern in watch_config[repo]['patterns']:
+                        if fnmatch(modified_file, pattern):
+                            notify = False
+                            break
```
Is this the right logic?  It amounts to, "Don't notify if any interesting file in this commit was modified in a previous commit for this pull request."   Shouldn't it be, "Don't notify if there's no interesting file in this commit that wasn't modified in a previous commit"?

lgp171188 replied:
@nedbat, the idea was to notify for the first interesting change in a PR and let people follow the PR for future updates. However, this doesn't work well when multiple unrelated watch patterns are added for a repository and a notification is sent only for the first match and not for the subsequent matches for the other patterns which might be interesting for a different set of users.

`Don't notify if there's no interesting file in this commit that wasn't modified in a previous commit` approach will work for the scenario mentioned above at the cost of more email notifications, which is something we wanted to keep to a minimum.

The app treats all the watch patterns for a repository as a single entity (the PR either matches the patterns or not, we don't care about which patterns were matched) and that is why it is not possible to do anything different than what we had in mind at the time of writing this.

However, the concerns you have raised are valid and we are open to addressing them.

CC @pomegranited
